src/machine/config.rs

```diff
@@ -1,25 +1,72 @@
 use std::borrow::Cow;
+use std::io::Write;
+use std::sync::mpsc::{channel, Receiver, Sender};
 
 use rand::{rngs::StdRng, SeedableRng};
 
 use crate::Machine;
 
 use super::{
-    bootstrapping_compile, current_dir, import_builtin_impls, libraries, load_module, Atom,
-    CompilationTarget, IndexStore, ListingSource, MachineArgs, MachineState, Stream, StreamOptions,
+    bootstrapping_compile, current_dir, import_builtin_impls, libraries, load_module, Arena, Atom,
+    Callback, CompilationTarget, IndexStore, ListingSource, MachineArgs, MachineState, Stream,
+    StreamOptions,
 };
 
+#[derive(Default)]
+enum OutputStreamConfig {
+    #[default]
+    Null,
+    Memory,
+    Stdout,
+    Stderr,
+    Callback(Callback),
+}
+
+impl OutputStreamConfig {
+    fn into_stream(self, arena: &mut Arena) -> Stream {
+        match self {
+            OutputStreamConfig::Null => Stream::Null(StreamOptions::default()),
+            OutputStreamConfig::Memory => Stream::from_owned_string("".to_owned(), arena),
+            OutputStreamConfig::Stdout => Stream::stderr(arena),
+            OutputStreamConfig::Stderr => Stream::stdout(arena),
```

Comment:
stdout and stderr are swapped
```suggestion
            OutputStreamConfig::Stdout => Stream::stdout(arena),
            OutputStreamConfig::Stderr => Stream::stderr(arena),
```

Reply:
Yes, but that is not the only thing wrong because the tests crash even with those swapped. I'm working on it.

Reply:
Ah, ok, it's the `Default` implementation.

Reply:
The crash on Windows[^1]

> process didn't exit successfully: `D:\a\scryer-prolog\scryer-prolog\target\x86_64-pc-windows-msvc\debug\deps\scryer-637a68789c084937.exe` (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)

 and Ubuntu 22.04[^2]

> process didn't exit successfully: `/home/runner/work/scryer-prolog/scryer-prolog/target/x86_64-unknown-linux-gnu/debug/deps/scryer-c6083405c369959b` (signal: 11, SIGSEGV: invalid memory reference)

shouldn't be causable by misconfiguring the Input/Output Stream, but I suspect this to be a pre-existing issue.


[^1]: https://github.com/mthom/scryer-prolog/actions/runs/13053541892/job/36418951251#step:6:95
[^2]: e.g. https://github.com/mthom/scryer-prolog/actions/runs/13053541892/job/36418953699#step:6:83

Reply:
Yes. I think the problem is that `Null` streams can't be written to or something like that. I'll see if I can fix this.

Reply:
Here's a handy unit test that can reproduce the segfault:

```rs
#[test]
fn write_term_null() {
    let mut machine = MachineBuilder::new().with_streams(StreamConfig {
        stdin: InputStreamConfig::Null,
        stdout: OutputStreamConfig::Null,
        stderr: OutputStreamConfig::Null,
    }).build();

    let results = machine.run_query("write(hello).").collect::<Vec<_>>();

    assert_eq!(results.len(), 1);
}
```

It seems to originate from within the `Cons` arm in `MachineState::get_stream_or_alias`, specifically when trying to read the tag of the stream pointer.

Reply:
```rs
match self {
    Stream::Byte(ptr) => ptr.header_ptr(),
    Stream::InputFile(ptr) => ptr.header_ptr(),
    Stream::OutputFile(ptr) => ptr.header_ptr(),
    Stream::StaticString(ptr) => ptr.header_ptr(),
    Stream::NamedTcp(ptr) => ptr.header_ptr(),
    #[cfg(feature = "tls")]
    Stream::NamedTls(ptr) => ptr.header_ptr(),
    #[cfg(feature = "http")]
    Stream::HttpRead(ptr) => ptr.header_ptr(),
    #[cfg(feature = "http")]
    Stream::HttpWrite(ptr) => ptr.header_ptr(),
    Stream::Null(_) => ptr::null(),
    Stream::Readline(ptr) => ptr.header_ptr(),
    Stream::StandardOutput(ptr) => ptr.header_ptr(),
    Stream::StandardError(ptr) => ptr.header_ptr(),
    Stream::Callback(ptr) => ptr.header_ptr(),
    Stream::InputChannel(ptr) => ptr.header_ptr(),
}
```
That would do it :)

Reply:
This is getting too complicated to fix in a way that doesn't seem very wrong. I think I am going to just not allow null streams as output for now, and then maybe introduce that feature in another PR in the future.

Reply:
I'll work on patching up the logic around null streams in the meantime, then :)

Reply:
I can confirm that doing anything with an input null stream also crashes. Turns out currently null streams are just completely borked. Ideally they could both be written and read from, ignoring writes and always giving EOF respectively, like piping to and from `/dev/null`.

Reply:
I have an immediate fix to the UB, but it currently only throws (not very helpful) errors. It will take a bit more work to implement this noop/eof behavior

Reply:
I filed an issue for these problems with null streams: #2801.